.github/workflows/main.yml

```diff
@@ -65,7 +65,113 @@ jobs:
       - name: Archive logs
         uses: actions/upload-artifact@v4
         with:
-          name: cxx-test-logs-macos
+          name: cxx-test-logs-linux
+          path: 'buildout/testlogs/*'
+          if-no-files-found: ignore
+        if: '!cancelled()'
+
+      - name: Teardown sccache
+        run: ./dev/ci/teardown-sccache
+        if: '!cancelled()'
+
+  cxx-linux-valgrind:
+    name: C++ tests on Valgrind
+    runs-on: ubuntu-24.04
+    timeout-minutes: 30
+    env:
+      USE_ASAN: false
+      EXTRA_CFLAGS: '-fdiagnostics-color'
+      EXTRA_CXXFLAGS: '-fdiagnostics-color'
+      SCCACHE_AZURE_BLOB_CONTAINER: ${{ vars.SCCACHE_AZURE_BLOB_CONTAINER }}
+      SCCACHE_AZURE_KEY_PREFIX: sccache/cxx-ubuntu-valgrind
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+
+      - uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: ${{ env.DEFAULT_RUBY_VERSION }}
+          bundler-cache: true
+
+      - name: Setup bundle gem path
+        run: ./dev/ci/setup-bundle-gem-path
+
+      - name: Setup sccache
+        run: ./dev/ci/setup-sccache
+        timeout-minutes: 1
+        env:
+          ARCH_AND_OS: x86_64-unknown-linux-musl
+          SCCACHE_AZURE_CONNECTION_STRING: ${{ secrets.AZURE_CI_STORAGE_CONNECTION_STRING }}
+
+      - name: Setup dependencies
+        run: >
+          sudo apt update &&
+          sudo apt install -y valgrind gdb libcurl4-openssl-dev
+
+      - name: Setup misc
+        run: |
+          set -x
+          sudo chmod 755 "$HOME"
+          cp test/config.json.github-ci-linux test/config.json
+
+      - name: Build C++ tests
+        run: bundle exec drake -j4 test:cxx:build
```

Comment:
don't use drake rake has -j flag

Reply:
Rake's -j doesn't work as you expect it to. It only works with multitask tasks which are tasks that always run in maximum concurrency in parallel and can't be controllable. Rake has no real equivalent to make's -j. You can test the following with `rake -j4` and see that it doesn't do what you want.

```ruby
task :default => [:t2, :t3, :t4] do
  puts "#{Time.now.to_f} t1"
end

task :t2 do
  puts "#{Time.now.to_f} t2"
  sleep 5
end

task :t3 do
  puts "#{Time.now.to_f} t3"
  sleep 5
end

task :t4 do
  puts "#{Time.now.to_f} t4"
  sleep 5
end
```

That's why we need drake.

Reply:
drake has been abandoned since 2011, I removed it because it was causing problems even back in 2021. I really don't see how relying on it is a good idea.

Reply:
What problems are you talking about? I'm still using it every day without issues. The CI has also been using it for months.

Furthermore, we don't have a good alternative. Compiling with standard rake is atrociously slow due to lack of concurrency. I'd rather for drake and fix any issues.

Reply:
Things seem to work for now. I'd say: let's keep using drake, then fork it once issues come up again.